src/prefer-filter-first.ts

```diff
@@ -0,0 +1,44 @@
+import { Rule } from "eslint";
+
+export const preferFilterFirstRule: Rule.RuleModule = {
+	meta: {
+		type: "suggestion",
+		docs: {
+			description:
+				"Calling `filter` first reduces the amount of iterations on the `map`.",
+			category: "Best Practices",
+			recommended: true,
+		},
+		// fixable: "code",
+	},
+	create: (context) => ({
+		CallExpression(node) {
+			if (node.callee.type !== "MemberExpression") {
+				return;
+			}
+
+			const callee = node.callee;
+			if (
+				callee.property.type !== "Identifier" ||
+				callee.property.name !== "map"
+			) {
+				return;
+			}
+
+			const parent = node.parent;
+			if (
+				parent.type !== "MemberExpression" ||
+				parent.property.type !== "Identifier" ||
+				parent.property.name !== "filter"
+			) {
+				return;
+			}
+
+			context.report({
+				node: parent,
+				message:
+					"Prefer using `arr.filter(a => !a).map(a => a)` instead of `arr.map(a => a).filter(a => !a)` to reduce the iterations the `map` runs over.",
```

Comment:
The endgame here is also to guide people towards using flatMap, as you can perform filter and map in the same pass. For example
```
[1,2,3].flatMap(i => i !== 2 ? i : [])
[1, 3]
```

Reply:
Great point! I think using https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/rules/prefer-array-flat-map.js makes even more sense now, because we could extend that rule to detect any usage of `arr`, filter, map, sort, (...), in any order and suggest `flatMap` instead.

Although I'm not sure if `flatMap` is faster, because it doesn't only allow filter + map, but also does a flattening step. Have you benchmarked it somewhere?

Reply:
I will write a simple benchmark, but this seems like a simple optimization that the engine could make by keeping track of the return value types and skip the flattening step altogether

Reply:
https://esbench.com/bench/65ddbc337ff73700a4debfd4 <-- assumption is correct! I thought maybe allocating [] could become expensive, but looks like it does not.